src/vm/internal/otable.cc

```diff
@@ -1,271 +1,131 @@
-      DCALLOC(otable_size, sizeof(object_t *), TAG_OBJ_TBL, "init_ch_otable"));
+ObjectTable::ObjectTable()
+:objects_({}),children_({})
+{}
 
-  for (x = 0; x < otable_size; x++) {
-    obj_table[x] = 0;
-    ch_table[x] = 0;
-  }
+//static method to return a pointer to the singleton object table.
+ObjectTable::Singleton ObjectTable::get() {
+    if( instance_ == nullptr )
+        instance_ = Singleton( new ObjectTable() );
+    return instance_;
 }
 
-/*
- * Looks for obj in table, moves it to head.
- */
-
-static long obj_searches = 0, obj_probes = 0, objs_found = 0;
-
-/* A global.  *shhhh* don't tell. */
-static int h;
-static int ch;
-
-static object_t *find_obj_n(const char *s) {
-  object_t *curr, *prev;
-  h = ObjHash(s);
-  curr = obj_table[h];
-  prev = 0;
-
-  obj_searches++;
-
-  while (curr) {
-    obj_probes++;
-    if (!strcmp(curr->obname, s)) { /* found it */
-      if (prev) {                   /* not at head of list */
-        prev->next_hash = curr->next_hash;
-        curr->next_hash = obj_table[h];
-        obj_table[h] = curr;
-      }
-      objs_found++;
-      return (curr); /* pointer to object */
+//fix style and efficiency errors in this function
+//should pass in a K const& k. also why is V not prefixed by ObjectTable and
+//it still works? is the ObjectTable needed in the parameter list? 
+
+//attempt to insert an key(obname), object pointer pair into the object table if the key is not in the table.
+//also will insert the key, object pointer pair in the childrens list.
+//if the key is already in the object table it fails and returns false, otherwise true.
+bool ObjectTable::insert(Key const& key,Value value) {
+    if( objects_.find(key) == objects_.end() ) {
+        objects_.insert( std::make_pair(key,value) );
+        auto base = basename(key);
+        auto it = children_.find(base);
+        if( it == children_.end() ) {
+          children_.insert( std::pair<Key,Vector>(base, {value} ) );
+        }
+        else {
+            it->second.push_back(value);
+        }
+        return true;
+    } 
+    else {
+        return false;
     }
-    prev = curr;
-    curr = curr->next_hash;
-  }
 
-  return (0); /* not found */
 }
 
-array_t *children(const char *s) {
-  auto max_array_size = CONFIG_INT(__MAX_ARRAY_SIZE__);
-
-  object_t *curr;
-  array_t *vec;
-  int size;
-  int count = 0;
-
-  s = _basename(s, &size);
-  ch = ObjHash(s);
-  curr = ch_table[ch];
-
-  vec = allocate_empty_array(max_array_size);
-  while (curr && count < max_array_size) {
-    if (!strncmp(curr->obname, s, size)) { /* found one */
-      vec->item[count].u.ob = curr;
-      add_ref(curr, "children");
-      vec->item[count].type = T_OBJECT;
-      count++;
+//attempt to find the object named key in the object table. Return a pointer to it, or if not found the null pointer.
+ObjectTable::Value ObjectTable::find(Key const & key) {
+    auto it = objects_.find(key);
+    if( it != objects_.end() ) {
```

Comment:
reverse the case here,  always put the error case first , so you don't have to write a "else{}"